core/src/test/java/io/confluent/kafka/schemaregistry/rest/RestApiCompatibilityTest.java

```diff
@@ -213,4 +225,234 @@ public void testCompatibilityLevelChangeToBackward() throws Exception {
             expectedIdSchema4,
             restApp.restClient.registerSchema(schemaString4, subject));
   }
+
+  @Test
+  public void testCompatibilityGroup() throws Exception {
+    String subject = "testSubject";
+
+    ParsedSchema schema1 = AvroUtils.parseSchema("{\"type\":\"record\","
+        + "\"name\":\"myrecord\","
+        + "\"fields\":"
+        + "[{\"type\":\"string\",\"name\":\"f1\"}]}");
+
+
+    ConfigUpdateRequest config = new ConfigUpdateRequest();
+    config.setCompatibilityGroup("application.version");
+    // add compatibility group
+    assertEquals("Adding compatibility group should succeed",
+        config,
+        restApp.restClient.updateConfig(config, null));
+
+    Map<String, String> properties = new HashMap<>();
+    properties.put("application.version", "1");
+    Metadata metadata1 = new Metadata(null, properties, null);
+    RegisterSchemaRequest request1 = new RegisterSchemaRequest(schema1);
+    request1.setMetadata(metadata1);
+    int expectedIdSchema1 = 1;
+    assertEquals("Registering should succeed",
+        expectedIdSchema1,
+        restApp.restClient.registerSchema(request1, subject, false));
+    // verify that default compatibility level is backward
+    assertEquals("Default compatibility level should be backward",
+        CompatibilityLevel.BACKWARD.name,
+        restApp.restClient.getConfig(null).getCompatibilityLevel());
+
+    // register forward compatible schema, which should fail
+    ParsedSchema schema2 = AvroUtils.parseSchema("{\"type\":\"record\","
+        + "\"name\":\"myrecord\","
+        + "\"fields\":"
+        + "[{\"type\":\"string\",\"name\":\"f1\"},"
+        + " {\"type\":\"string\",\"name\":\"f2\"}]}");
+    RegisterSchemaRequest request2 = new RegisterSchemaRequest(schema2);
+    try {
+      restApp.restClient.registerSchema(request2, subject, false);
+      fail("Registering a forward compatible schema should fail");
+    } catch (RestClientException e) {
+      // this is expected.
+      assertEquals("Should get a conflict status",
+          RestIncompatibleSchemaException.DEFAULT_ERROR_CODE,
+          e.getStatus());
+    }
+
+    // now try registering a forward compatible schema in a different compatibility group
+    properties = new HashMap<>();
+    properties.put("application.version", "2");
+    Metadata metadata2 = new Metadata(null, properties, null);
+    request2.setMetadata(metadata2);
+    int expectedIdSchema2 = 2;
+    assertEquals("Registering should succeed",
+        expectedIdSchema2,
+        restApp.restClient.registerSchema(request2, subject, false));
+  }
+
+  @Test
+  public void testConfigMetadata() throws Exception {
```

Comment:
I'm trying to understand what is the `finalInitialMetadata`. From this test, the `initialMetadata` is the default for all subject, and passed metadata is the subject-version-specific metadata so it won't be propagate to the next version, is that correct?

Reply:
The metadata/ruleset is calculated as follows:

1.  If the initial metadata/ruleset exists in the config, start with that, call it X.
2. Get the specific metadata/ruleset.  This is either the one that's passed in with the schema, or if it's null in the passed in schema, then use the one in the previous schema
3. Merge the specific metadata/ruleset on top of X.
4. Merge the final metadata/ruleset from the config (if it exists) on top of X.
5. X is now the value of the metadata/ruleset that will be saved with the schema.